### ISSUE TITLE
Analysis of Ignored Patches in the Linux Kernel Development

### ISSUE THREAD

**Issue by hsanchez** [hsanchez/paper-memory#3]
## TL;DR

This Master thesis reports the results of statistical analyses of the ignored patch phenomenon in the Linux Kernel development.

### Paper Link

[osr.cs.fau.de/../duda_2019](https://osr.cs.fau.de/wp-content/uploads/2020/03/duda_2019.pdf)

### Author/Institution

- Sebastian Duda (riedrich-Alexander-Universität Erlangen-Nürnberg)

## Overview

This thesis reports the results of a statistical analysis performed on the Linux Kernel over a time frame to study the ignore patches phenomenon. In the analyzed time frame (release v3.0 to v4.20), 18k of 792k patches (2.3%) were ignored. The ratio of ignored patches also decreased over time. Two clusters (minor contributions, and automatically created patches) of patches making up most of the ignored patches were detected. There were no other statistically significant abnormalities of the ignored patches. However, based on the analysis, the authors showed that there were indicators of discrimination against certain groups. The author further recognized a trend that **larger subsystems and lists were ignoring relatively fewer patches**.

## Contributions and Distinctions from Previous Works

**Contributions:**

- An extensive and sophisticated dataset of the patches submitted between the 19th of May 2011 (release of Linux kernel version 2.6.39) and the 23rd of November 2018 (release of kernel version 4.20).
- A guide for how this dataset can be used to generate a dataset of the authors/mailing lists/subsystems for further analysis.
- Two spot tests to validate all measurements in terms of the exactness and reliability of the data.
- A definition of when a patch is ignored.

**Difference:**

The Linux kernel development process is well understood, documented, and researched. However, some actions of contributors or maintainers do not comply with the process. One of these actions is ignoring patches. Ignoring a patch
means that the patch is neither answered nor accepted by the developers. There is no analysis of this phenomenon yet.
This work is the first one to provide such an analysis.

## Methods

- An exploratory, embedded, single-case case study with two stages: (1) data collection, (2) data analysis.
- Selection of measurements. The measurements were collected in an iterative two-step process: (1) brainstorming of new measurements, (2)  discussion of the ideas in a peer group.
- Exploratory data analysis. Used John W. Tukey’s 5-number-summary and the suggested sqrt-scale in some plots.

## Results

- In the measured time frame (v3.0 to v4.20), there were 18k patches of 792k patches ignored; about 2, 3%.
- Found no characteristic of ignored patches that is statistically sound.
- Some authors are ignored more often due to the type of patches they sent.
- Ignored patches are equally distributed, independent of specific characteristics of the mailing lists.
- Detected the trend that large subsystems (by patch frequency) are ignoring fewer patches than smaller subsystems.
- Ignored patches can be grouped into two clusters: automatically created, and minor contributions.
